### PR TITLE
Add more scenerios to win virtio driver test

### PR DESCRIPTION
**Pull request by suqinhuang**
1. Add test before/after/during driver update
2. Add viorng load/unload test

### CODE REVIEW

**Comment by suqinhuang** [comment]
@CongLi could you review this patch

**Comment by suqinhuang** [comment]
id: 1358669

File: qemu/tests/cfg/win_virtio_driver_update_test.cfg [hunk 1]
```diff
@@ -87,9 +89,26 @@
...
         - uninstall_install:
             need_uninstall = yes
         - downgrade_upgrade:
             need_uninstall = no
             cdrom_virtio_downgrade = isos/windows/virtio-win-downgrade.iso
+    variants:
+        - before_bg_test:
+            run_bg_flag = "before_bg_test"
+        - during_bg_test:
+            run_bg_flag = "during_bg_test"
```
Comment (CongLi) [inline qemu/tests/cfg/win_virtio_driver_update_test.cfg]: Need ignore error during driver test.


15:39:00 ERROR| ShellProcessTerminatedError: Shell process terminated while waiting for command to complete: 'echo list disk > disk && echo exit >> disk && diskpart /s disk'    (status: 1,    output: '\nMicrosoft DiskPart version 6.3.9600\n\nCopyright (C) 1999-2013 Microsoft Corporation.\nOn computer: WIN-IE24DTPILE3\n\nC:\\>Ncat: Connection reset by peer.\n')

**Comment by suqinhuang** [comment]
PASS 1-smp_1.4096m.repeat1.run_test.Host_RHEL.m7.u2.rng_random.qcow2.virtio_blk.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.win_virtio_driver_update_test.before_bg_test.downgrade_upgrade.with_viorng

PASS 1-smp_1.4096m.repeat1.run_test.Host_RHEL.m7.u2.rng_random.qcow2.virtio_blk.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.win_virtio_driver_update_test.after_bg_test.downgrade_upgrade.with_viorng

**Comment by suqinhuang** [comment]
during_bg_test is removed, as it's not requested

**Comment by suqinhuang** [comment]
@CongLi could you help review this patch

File: qemu/tests/cfg/win_virtio_driver_update_test.cfg [hunk 2]
```diff
@@ -87,9 +90,27 @@
...
             iozone_timeout = 7200
             wait_bg_time = 180
+        - with_viorng:
+            no no_virtio_rng
+            driver_name = viorng
+            device_name = "VirtIO RNG Device"
+            run_bgstress = rng_bat 
+            target_process = random\w*.exe
+            session_cmd_timeout = 240
+            read_rng_cmd  = "X:\random_%PROCESSOR_ARCHITECTURE%.exe"
```
Comment (CongLi) [inline qemu/tests/cfg/win_virtio_driver_update_test.cfg]: Hi @suqinhuang 

Same problem of rng for windows guest.

2016-08-18 14:07:13: C:\Users\Administrator>
2016-08-18 14:07:13: D:\random_%PROCESSOR_ARCHITECTURE%.exe
2016-08-18 14:07:13: Generated random number is
2016-08-18 14:07:13: 0x7f, 0x97, 0x0d, 0x42, 0x57, 0x61, 0x72, 0xd1, 0x00, 0x8f,
2016-08-18 14:07:13: 0x83, 0x42, 0xf5, 0x16, 0x10, 0x0b, 0x21, 0x6c, 0x82, 0x43,
2016-08-18 14:07:13: 0xe3, 0x8a, 0x1e, 0x7f, 0x9e, 0xd9, 0xfb, 0x8d, 0x1a, 0x37,
2016-08-18 14:07:13: 0x77, 0x82, 0x77, 0x55, 0x9c, 0x1c, 0x11, 0xe7, 0x21, 0x0a,
2016-08-18 14:07:13: 0x7e, 0xa5, 0xe8, 0x82, 0x13, 0xc8, 0x77, 0x86, 0x2f, 0x41,
2016-08-18 14:07:13: 0xb2, 0x63, 0x8b, 0x1d, 0x51, 0x33, 0x44, 0xcb, 0xdb, 0x4c,
2016-08-18 14:07:13: 0xbb, 0x8e, 0x01, 0x8d, 0xf4, 0xdc, 0x13, 0x80, 0x79, 0xc7,
2016-08-18 14:07:13: 0x10, 0x03, 0x7b, 0xc5, 0xba, 0x0d, 0x53, 0x32, 0x39, 0x1b,
2016-08-18 14:07:13: 0x98, 0x5c, 0xd3, 0xef, 0xcd, 0xa7, 0xf3, 0xfa, 0xc0, 0x15,
2016-08-18 14:07:13: 0x85, 0x93, 0x40, 0x88, 0xcb, 0x78, 0x8e, 0x0b, 0x9c, 0x52,
2016-08-18 14:07:13: 0xf4, 0x09, 0x8d, 0xff, 0x1e, 0x82, 0xbe, 0x0f, 0x06, 0xc4,
2016-08-18 14:07:13: 0x38, 0x4a, 0x48, 0x3b, 0x13, 0x0e, 0x7c, 0xcb, 0xc6, 0xa0,
2016-08-18 14:07:13: 0x86, 0x0c, 0xf9, 0xe9, 0x2a, 0xa0, 0xcd, 0x0b,

Time if too short to check the rng generation process for Win2012r2 guest.
14:09:17 ERROR| FAIL 1-smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_blk.up.virtio_net.Win2012.x86_64.r2.io-github-autotest-qemu.win_virtio_driver_update_test.during_bg_test.uninstall_install.with_viorng -> TestFail: Backgroud test rng_bat is not alive!

**Comment by PyLearner** [comment]
got error 

04:09:09 ERROR| Reproduced traceback from: /usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/avocado_vt/test.py:420
04:09:09 ERROR| Traceback (most recent call last):
04:09:09 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/avocado_vt/test.py", line 196, in runTest
04:09:09 ERROR|     self._runTest()
04:09:09 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/avocado_vt/test.py", line 337, in _runTest
04:09:09 ERROR|     run_func(self, params, env)
04:09:09 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/virttest/error_context.py", line 135, in new_fn
04:09:09 ERROR|     return fn(*args, **kwargs)
04:09:09 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/driver_in_use.py", line 104, in run
04:09:09 ERROR|     run_subtest(bg_stress_test)
04:09:09 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/driver_in_use.py", line 80, in run_subtest
04:09:09 ERROR|     utils_test.run_virt_sub_test(test, params, env, sub_type)
04:09:09 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/virttest/utils_test/__init__.py", line 1242, in run_virt_sub_test
04:09:09 ERROR|     run_func(test, params, env)
04:09:09 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/virttest/error_context.py", line 135, in new_fn
04:09:09 ERROR|     return fn(*args, **kwargs)
04:09:09 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/rng_bat.py", line 100, in run
04:09:09 ERROR|     "guest: %s" % output)
04:09:09 ERROR| TestFail: Unable to read random numbers fromguest: **** Error 0xc0000135 returned by BCryptOpenAlgorithmProvider
04:09:09 ERROR|
04:09:09 ERROR|
04:09:09 ERROR|
04:09:09 ERROR| FAIL 1-smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_blk.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.win_virtio_driver_update_test.before_bg_test.downgrade_upgrade.with_viorng -> TestFail: Unable to read random numbers fromguest: **** Error 0xc0000135 returned by BCryptOpenAlgorithmProvider

**Comment by PyLearner** [comment]
Acked-by : Meng Yang meyang@redhat.com

**Comment by pingl2016** [comment]
Case win_virtio_driver_update_test.during_bg_test.uninstall_install.with_viorng and win_virtio_driver_update_test.after_bg_test.uninstall_install.with_viorng failed.
03:01:47 ERROR| Reproduced traceback from: /usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/avocado_vt/test.py:420
03:01:47 ERROR| Traceback (most recent call last):
03:01:47 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/avocado_vt/test.py", line 196, in runTest
03:01:47 ERROR|     self._runTest()
03:01:47 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/avocado_vt/test.py", line 337, in _runTest
03:01:47 ERROR|     run_func(self, params, env)
03:01:47 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/virttest/error_context.py", line 135, in new_fn
03:01:47 ERROR|     return fn(*args, **kwargs)
03:01:47 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/driver_in_use.py", line 119, in run
03:01:47 ERROR|     run_subtest(sub_type)
03:01:47 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/driver_in_use.py", line 80, in run_subtest
03:01:47 ERROR|     utils_test.run_virt_sub_test(test, params, env, sub_type)
03:01:47 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/virttest/utils_test/__init__.py", line 1242, in run_virt_sub_test
03:01:47 ERROR|     run_func(test, params, env)
03:01:47 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/virttest/error_context.py", line 135, in new_fn
03:01:47 ERROR|     return fn(*args, **kwargs)
03:01:47 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/win_virtio_driver_update_test.py", line 33, in run
03:01:47 ERROR|     single_driver_install.run(test, params, env)
03:01:47 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/virttest/error_context.py", line 135, in new_fn
03:01:47 ERROR|     return fn(*args, **kwargs)
03:01:47 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/single_driver_install.py", line 133, in run
03:01:47 ERROR|     session = reboot(vm, session)
03:01:47 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/single_driver_install.py", line 62, in reboot
03:01:47 ERROR|     return vm.reboot(nic_index=nic_idx)
03:01:47 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/virttest/error_context.py", line 135, in new_fn
03:01:47 ERROR|     return fn(*args, **kwargs)
03:01:47 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/virttest/qemu_vm.py", line 3914, in reboot
03:01:47 ERROR|     shell_reboot(session, nic_index, serial, timeout / 2)
03:01:47 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/virttest/qemu_vm.py", line 3901, in shell_reboot
03:01:47 ERROR|     session = __get_session(nic_index=nic_index, serial=serial)
03:01:47 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/virttest/qemu_vm.py", line 3854, in __get_session
03:01:47 ERROR|     return self.login(nic_index=nic_index)
03:01:47 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/virttest/error_context.py", line 135, in new_fn
03:01:47 ERROR|     return fn(*args, **kwargs)
03:01:47 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/virttest/virt_vm.py", line 944, in login
03:01:47 ERROR|     interface=neigh_attach_if)
03:01:47 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/virttest/remote.py", line 254, in remote_login
03:01:47 ERROR|     handle_prompts(session, username, password, prompt, timeout)
03:01:47 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/virttest/remote.py", line 174, in handle_prompts
03:01:47 ERROR|     raise LoginError("Client said 'connection closed'", text)
03:01:47 ERROR| LoginError: Client said 'connection closed'    (output: '\nDenying new connections due to the limit on number of connections.\x00Connection closed by foreign host.\n')
03:01:47 ERROR|
03:01:47 ERROR| Traceback (most recent call last):

03:01:47 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/avocado_vt/test.py", line 434, in _run_avocado
    raise test_exception

03:01:47 ERROR| LoginError: Client said 'connection closed'    (output: '\nDenying new connections due to the limit on number of connections.\x00Connection closed by foreign host.\n')

03:01:47 ERROR| ERROR 3-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.win_virtio_driver_update_test.after_bg_test.uninstall_install.with_viorng -> LoginError: Client said 'connection closed'    (output: '\nDenying new connections due to the limit on number of connections.\x00Connection closed by foreign host.\n')

**Comment by pingl2016** [comment]
Test with below patch, all the cases passed as expected.
(1/7) smp_1.2048m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.win_virtio_driver_update_test.with_viorng.before_bg_test.uninstall_install:  PASS (583.65 s)
 (2/7) smp_1.2048m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.win_virtio_driver_update_test.with_viorng.before_bg_test.downgrade_upgrade:  PASS (656.65 s)
 (3/7) smp_1.2048m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.win_virtio_driver_update_test.with_viorng.during_bg_test.uninstall_install:  PASS (466.17 s)
 (4/7) smp_1.2048m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.win_virtio_driver_update_test.with_viorng.during_bg_test.downgrade_upgrade:  PASS (488.77 s)
 (5/7) smp_1.2048m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.win_virtio_driver_update_test.with_viorng.after_bg_test.uninstall_install:  PASS (556.58 s)
 (6/7) smp_1.2048m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.win_virtio_driver_update_test.with_viorng.after_bg_test.downgrade_upgrade:  PASS (584.33 s)

diff --git a/qemu/tests/single_driver_install.py b/qemu/tests/single_driver_inst
index 2a113ff..41cb7d1 100644
--- a/qemu/tests/single_driver_install.py
+++ b/qemu/tests/single_driver_install.py
@@ -130,6 +130,7 @@ def run(test, params, env):
     try:
         error_context.context("Show file extentions", logging.info)
         show_file_extentions(session)
+        session.close()
         session = reboot(vm, session)
 
         uninstall_flag = params.get("need_uninstall", "no")

**Comment by suqinhuang** [comment]
new issue is created to trace the connection limitation in win7 guest

**Comment by suqinhuang** [comment]
the connection issue will be fixed with a separate patch, merge this patch first.

**Pull request merged**